src/NuGet.Core/NuGet.Versioning/NuGetVersionFactory.cs

```diff
@@ -112,6 +104,98 @@ int IndexOf(string str, char c)
+                    GetNextSection(str, endIndex + 1, out endIndex, out build) &&
+                    GetNextSection(str, endIndex + 1, out endIndex, out revision) &&
+                    endIndex == str.Length;
+
+                if (validString)
+                {
+                    version = new Version(major, Math.Max(0, minor), Math.Max(0, build), Math.Max(0, revision));
+                    return true;
+                }
+            }
+
+            version = null;
+
+            return false;
+
+            // returns false if an invalid section was found while processing the string
+            static bool GetNextSection(string s, int start, out int end, out int versionNumber)
+            {
+                // check to see if we've processed the whole string
+                if (start >= s.Length)
+                {
+                    // we've reached the end. The section is empty but not invalid.
+                    end = s.Length;
+                    versionNumber = -1;
+
+                    return true;
+                }
+
+                end = s.IndexOf('.', start);
+
+                if (end == -1)
+                {
+                    end = s.Length;
+                }
+
+                return TryParseInt(s, start, end, out versionNumber);
+            }
+
+            // start is inclusive bound and end is exclusive for the section of string to parse
+            static bool TryParseInt(string s, int start, int end, out int value)
+            {
+                // invalid section specified
+                if (start >= end)
+                {
+                    value = 0;
+                    return false;
+                }
+
+                long returnValue = 0;
+
+                // Parse the values in reverse and assign a tens place to them.
+                // When parsing "123456", this method adds 6 + 50 + 400 + 3000 + 20000 + 100000
+                for (int i = start; i < end; i++)
+                {
+                    // negative numbers are invalid for version strings so we only need to check for digits
+                    char current = s[i];
+                    if (current < '0' || current > '9')
+                    {
+                        value = 0;
+                        return false;
+                    }
+
+                    // subtract off ASCII value of '0' from our current character to get the digit's value
+                    // e.g. '3' - '0' == 51 - 58 == 3
+                    returnValue = returnValue * 10 + (current - '0');
+
+                    // Check for overflow. We can't get outside the bounds of long before exceeding int.MaxValue
+                    // Intentionally avoid usage of 'checked' statement to avoid exception
+                    if (returnValue > int.MaxValue)
+                    {
+                        value = 0;
+                        return false;
+                    }
```

Comment:
@jeffkl and @zivkan, I really want to thank you both for spending time helping me ensure that this code is up to snuff 😄. I took a step back and rewrote my approach to address some of the issues we found (incorporating some things I picked up from @jeffkl 's implementation) and added a battery of tests (over 100) to test all the cases that were brought up and any additional ones I can think of.

I admit, I wasn't anticipating that `Version.Parse` and `Version.TryParse` handled so many different cases when I started this. The documentation for `Version.Pars` is pretty light on details 😄 

"The input parameter must have the following format:

major.minor[.build[.revision]]

where major, minor, build, and revision are the string representations of the version number's four components: major version number, minor version number, build number, and revision number, respectively. Optional components are shown in square brackets ([ and ]). The components must appear in the specified order and must be separated by periods."

@zivkan, I'm not entirely opposed to using `checked`, but there is a tiny bit of overhead even in the good case (super small). My bigger concern is around the error case since exceptions are expensive and methods with the BCL `Try*` methods don't throw at all. Unless there's a good reason to, I think we should follow that pattern.

I'd be curious to see what test code you used with `10000000000000000000`. The calculations are done using `long`, so seeing any overflow seem impossible to me since the code terminates as soon as we exceed `int.MaxValue`.